infoschema/builder.go

```diff
@@ -173,21 +153,35 @@ func (b *Builder) applyCreateTable(m *meta.Meta, roDBInfo *model.DBInfo, tableID
 	sortedTables = append(sortedTables, tbl)
 	sort.Sort(sortedTables)
 	b.is.sortedTablesBuckets[bucketIdx] = sortedTables
+
+	newTbl, ok := b.is.TableByID(tableID)
+	if ok {
+		roDBInfo.Tables = append(roDBInfo.Tables, newTbl.Meta())
+	}
 	return nil
 }
 
-func (b *Builder) applyDropTable(di *model.DBInfo, tableID int64) {
+func (b *Builder) applyDropTable(roDBInfo *model.DBInfo, tableID int64) {
 	bucketIdx := tableBucketIdx(tableID)
 	sortedTables := b.is.sortedTablesBuckets[bucketIdx]
 	idx := sortedTables.searchTable(tableID)
 	if idx == -1 {
 		return
 	}
-	if tableNames, ok := b.is.schemaMap[di.Name.L]; ok {
+	if tableNames, ok := b.is.schemaMap[roDBInfo.Name.L]; ok {
 		delete(tableNames.tables, sortedTables[idx].Meta().Name.L)
 	}
 	// Remove the table in sorted table slice.
 	b.is.sortedTablesBuckets[bucketIdx] = append(sortedTables[0:idx], sortedTables[idx+1:]...)
+
+	// The old DBInfo still holds a reference to old table info, we need to remove it.
+	tables := make([]*model.TableInfo, 0, len(roDBInfo.Tables)-1)
+	for _, tblInfo := range roDBInfo.Tables {
```

Comment:
How about find the old table's position first and then roDBInfo.Tables =  append(roDBInfo.Tables[:pos], roDBInfo.Tables[pos+1:])